core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/io/ChecksumSubscriber.java

```diff
@@ -18,78 +18,52 @@
-import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import java.util.zip.Checksum;
 import org.reactivestreams.Subscriber;
-import org.reactivestreams.Subscription;
 import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.async.DelegatingSubscriber;
 
 /**
  * A subscriber that takes a collection of checksums, and updates each checksum when it receives data.
  */
 @SdkInternalApi
-public final class ChecksumSubscriber implements Subscriber<ByteBuffer> {
-    private final CompletableFuture<Void> checksumming = new CompletableFuture<>();
+public final class ChecksumSubscriber extends DelegatingSubscriber<ByteBuffer, ByteBuffer> {
+    private final CompletableFuture<Void> signal;
     private final Collection<Checksum> checksums = new ArrayList<>();
-    private volatile boolean canceled = false;
-    private volatile Subscription subscription;
 
-    public ChecksumSubscriber(Collection<? extends Checksum> consumers) {
-        this.checksums.addAll(consumers);
-
-        checksumming.whenComplete((r, t) -> {
-            if (t instanceof CancellationException) {
-                synchronized (this) {
-                    canceled = true;
-                    if (subscription != null) {
-                        subscription.cancel();
-                    }
-                }
-            }
-        });
-    }
+    public ChecksumSubscriber(Subscriber<? super ByteBuffer> subscriber, Collection<? extends Checksum> consumers,
+                              CompletableFuture<Void> signal) {
+        super(subscriber);
 
-    @Override
-    public void onSubscribe(Subscription subscription) {
-        synchronized (this) {
-            if (!canceled && this.subscription == null) {
-                this.subscription = subscription;
-                subscription.request(Long.MAX_VALUE);
-            } else {
-                subscription.cancel();
-            }
-        }
+        this.checksums.addAll(consumers);
+        this.signal = signal;
     }
 
     @Override
     public void onNext(ByteBuffer byteBuffer) {
-        if (!canceled) {
-            byte[] buf;
-
-            if (byteBuffer.hasArray()) {
-                buf = byteBuffer.array();
-            } else {
-                buf = new byte[byteBuffer.remaining()];
-                byteBuffer.get(buf);
-            }
-
-            // We have to use a byte[], since update(<ByteBuffer>) is java 9+
-            checksums.forEach(checksum -> checksum.update(buf, 0, buf.length));
+        byte[] buf;
+        if (byteBuffer.hasArray()) {
+            buf = byteBuffer.array();
+        } else {
+            buf = new byte[byteBuffer.remaining()];
+            byteBuffer.get(buf);
         }
+        // We have to use a byte[], since update(<ByteBuffer>) is java 9+
+        checksums.forEach(checksum -> checksum.update(buf, 0, buf.length));
+
+        subscriber.onNext(byteBuffer);
```

Comment:
Should we copy the default implementation of `update(ByteBuffer)` from Java 9+? It honors the bounds of the byte buffer properly. This implementation doesn't seem to reset the buffer when `byteBuffer.hasArray()` is false, either. Here's a version where I use `update(ByteBuffer)` from Java 9+ and reset the buffer after using it:

```java
@SdkInternalApi
public final class ChecksumSubscriber extends DelegatingSubscriber<ByteBuffer, ByteBuffer> {
    private final CompletableFuture<Void> signal;
    private final byte[] copyBuffer;
    private final Collection<Checksum> checksums = new ArrayList<>();

    public ChecksumSubscriber(Subscriber<? super ByteBuffer> subscriber, 
                              Collection<? extends Checksum> consumers,
                              CompletableFuture<Void> signal) {
        super(subscriber);

        this.checksums.addAll(consumers);
        this.signal = signal;
    }

    @Override
    public void onNext(ByteBuffer byteBuffer) {
        updateChecksumsAndReset(byteBuffer);
        subscriber.onNext(byteBuffer);
    }

    private void updateChecksumsAndReset(ByteBuffer buffer) {
        int position = buffer.position();
        int limit = buffer.limit();
        int remaining = limit - position;
        if (remaining <= 0) {
            return;
        }
        if (buffer.hasArray()) {
            checksums.forEach(c -> c.update(buffer.array(), position + buffer.arrayOffset(), remaining));
        } else {
            if (copyBuffer == null) {
                copyBuffer = new byte[4096];
            }
            while (buffer.hasRemaining()) {
                int length = Math.min(buffer.remaining(), copyBuffer.length);
                buffer.get(copyBuffer, 0, length);
                checksums.forEach(c -> c.update(copyBuffer, 0, length));
            }
        }
        buffer.position(position);
        buffer.limit(limit);
    }

    @Override
    public void onError(Throwable t) {
        super.onError(t);
        signal.completeExceptionally(t);
    }

    @Override
    public void onComplete() {
        super.onComplete();
        signal.complete(null);
    }
}
```